### ISSUE TITLE
[MNT]: Please update freetype version

### ISSUE THREAD

**Issue by stphnlyd** [matplotlib/matplotlib#20842]
### Summary

Now the library still by default uses freetype 2.6.1
https://github.com/matplotlib/matplotlib/blob/3a4fdea8d23207d67431973fe5df1811605c4132/setupext.py#L174
Freetype 2.6.1 is a very old version. Today latest freetype release is 2.11.0. And some freetype release like 2.10.4 has important vulnerability fix. See https://www.freetype.org/index.html#news

### Proposed fix

Support newer versions of freetype, and by default use its latest version that is 2.11.0

**Comment by mathstuf** [comment]
Note that upstream freetype has moved to CMake on Windows (`ftjam` is gone at least), so the build scripts will need updated here.

**Comment by QuLogic** [comment]
The vulnerability in 2.10.4 is in PNG support, which is not enabled in our FreeType build. All other published CVEs have been reviewed and are not applicable to FreeType as used in Matplotlib.

There are pros and cons to moving forward, but known vulnerabilities are not (currently) one of them.

**Comment by jklymak** [comment]
Is the "local" build really the default, or just a testing default?

**Comment by QuLogic** [comment]
It is the default; manylinux does not guarantee freetype is available on Linux and it isn't on macOS or Windows either.

**Comment by tacaswell** [comment]
The main reason we are pinned to 2.6.1 is that we do 0 tolerance image comparison tests, if we change the version of freetype, the text in the images changes a bit and breaks our tests (in the past we had a default non-zero tolerance and that let through several horrific bugs where we were rendering the wrong glyph but it still was "within tolerance" 🤦🏻 ).  We need to revive the GSOC work from last summer to improve the management of our baseline images before we change the freetype version we are pinned to.

Down-stream packagers can (and do at least in the case of Fedora and Conda) distribute newer versions of freetype and link us against that version.  See https://matplotlib.org/stable/devel/dependencies.html#c-libraries for instructions on how to link against your system freetype.

**Comment by ArchangeGabriel** [comment]
I would also reference https://github.com/QuLogic/mpl-images for people wanting to run tests against a more recent FreeType.

In #21271 the repo size was mentioned as an issue, have you considered using git-lfs to counter this?

**Comment by tacaswell** [comment]
> have you considered using git-lfs to counter this

iirc when we looked at the space / bandwidth you could get it was too small to be useful.  The problem is not that we have a few big files the problem is we have have a large number of small files.   If we were going to go a git-level solution I think sub-modules would be a better path.

However, a git-level solution does not get around the problem of needing to switch which version of the images you are using to match the freetype you are building with.  Even if we were to track and checkin re-built images with the new freetype we would likely still be broken for a majority of the down-stream packagers as we standardized on a different version than they ship (which is arguably better than now where we are broken for _everyone_ but the current case could be said to be more fair 😉 ).

**Comment by jklymak** [comment]
It seems the easiest is to just have an image repo and if folks are concerned about its size, they can just do a shallow clone.  The only problem with that is getting CI to work smoothly between the two repos in that you'd want the main PR to use the new or modified image in the image repo.

**Comment by jklymak** [comment]
Or maybe even try a submodule setup.  https://github.blog/2016-02-01-working-with-submodules/

**Comment by ConorMacBride** [comment]
> It seems the easiest is to just have an image repo and if folks are concerned about its size, they can just do a shallow clone. The only problem with that is getting CI to work smoothly between the two repos in that you'd want the main PR to use the new or modified image in the accompanying image repo pr

A URL to a remote repo can be specified in https://github.com/matplotlib/pytest-mpl for the baseline image directory. SunPy [uses this feature](https://github.com/sunpy/sunpy-figure-tests) and [updates the image repo](https://github.com/sunpy/sunpy/blob/b8444bb8a359912e51dd89fcf0246dd67bad718a/.circleci/config.yml#L55) when `main` is pushed to. Would it make sense for matplotlib to migrate to using the matplotlib/pytest-mpl plugin?

**Comment by jklymak** [comment]
> 
> A URL to a remote repo can be specified in https://github.com/matplotlib/pytest-mpl for the baseline image directory. SunPy [uses this feature](https://github.com/sunpy/sunpy-figure-tests) and [updates the image repo](https://github.com/sunpy/sunpy/blob/b8444bb8a359912e51dd89fcf0246dd67bad718a/.circleci/config.yml#L55) when `main` is pushed to. Would it make sense for matplotlib to migrate to using the matplotlib/pytest-mpl plugin?

That sounds good.  I assume a new PR can compare against its own new/modified image?

**Comment by ConorMacBride** [comment]
Yes, however for a *remote* baseline image repository it's best to use the [Hybrid Mode](https://github.com/matplotlib/pytest-mpl#hybrid-mode-hashes-and-images) so generated images are compared against sha hashes of the baseline image files, which would be stored in the `matplotlib/matplotlib` repo. This way you don't need to update the remote baseline image repository *before* the PR is merged in order for the tests to pass.

**Comment by jklymak** [comment]
@ConorMacBride Following up on this: what happens if someone writing a PR _changes_ an existing image or adds a new image?  Do they open two PRs, one in each repo?  Seems that CI wouldn't know where to find the new PR in the image repo?

**Comment by ConorMacBride** [comment]
@jklymak The image repo is only updated when the PR is merged into the main (matplotlib/matplotlib) branch, so only one PR in total is needed. This update can be configured to happen automatically as SunPy did [here](https://github.com/sunpy/sunpy/blob/b8444bb8a359912e51dd89fcf0246dd67bad718a/.circleci/config.yml#L55). This CI runs in the main repository and pushes the updates to the remote image repository.

When using the Hybrid Mode, only the updated/new image *hashes* need to be added to the main PR for the tests to pass. This is because actual images are only compared against the baseline images if the generated hashes do not match. (The hashes are sensitive to the particular OS and freetype version, but a hash library for each version can be included in the main repo, and the correct library is selected for the test environment.) I've been working on some tools (https://github.com/matplotlib/pytest-mpl/pull/128) to hopefully make it easier for contributors to update the hashes and see how their changes affect the baseline images.

**Comment by jklymak** [comment]
> to hopefully make it easier for contributors to update the hashes and see how their changes affect the baseline images.

Right, this seems key, because if I change a bunch of images, it is really helpful for there to be a way to browse the changes in the PR.  However, I guess if I generate an updated image, I would still make a PR on the image repo and folks could browse the diff there, even if the main-repo tests are run on the hash?

**Comment by ConorMacBride** [comment]
The manual PR to the image repo shouldn't be necessary if you can have the CI provider make the test artifacts (i.e. generated and diff images) available to download. SunPy uses CircleCI for the image comparison tests as it can serve HTML artifacts, so the difference images are all presented in a HTML report, [this for example](https://74223-2165383-gh.circle-artifacts.com/0/.tmp/py38-figure/figure_test_images/fig_comparison.html). Although I have a PR to present a report like [this](https://macbride.me/pytest-mpl/interactive-html/fig_comparison.html). This way the baseline images always correspond to the state of the main branch of the main repo.

**Comment by apteryks** [comment]
Could we have a pytest marker to disable all these freetype/pinned dependent tests? Otherwise the various distributions (which build using an up-to-date freetype) have to either disable the whole test suite or play whack-a-mole ignoring the affected tests.

**Comment by apteryks** [comment]
Found a relatively concise way to express the tests to be skipped, but would still be nice to have a marker:
```scheme
(define-public python-matplotlib
  (package
    (name "python-matplotlib")
    (version "3.10.8")
    (source (origin
              (method git-fetch)
              (uri (git-reference
                     (url "https://github.com/matplotlib/matplotlib")
                     (commit (string-append "v" version))))
              (file-name (git-file-name name version))
              (sha256
               (base32
                "1rlday52hcyrp7xfbdf4z8lzgnxkpfvjxnl9qmq0jvc3ph3n8k5i"))))
    (build-system pyproject-build-system)
    (arguments
     (list
      #:configure-flags
      ''(("setup-args" . #("-DrcParams-backend=TkAgg"
                           "-Dsystem-freetype=true"
                           "-Dsystem-qhull=true"))
         ("install-args" . "--tags=data,python-runtime,runtime,tests"))
      #:test-flags
      #~(list "-m" "not network"
              "-n" (number->string (parallel-job-count))
              ;; Run the installed tests, which is what we want since not
              ;; everything gets built in the source directory.
              "--pyargs" "matplotlib"
              "-k"
              (string-join
               (list
                ;; XXX: Disable all image comparison tests because we're using
                ;; a newer version of FreeType than matplotlib expects.  This
                ;; leads to minor differences throughout the tests.
                "not image_comparison"
                "check_figures_equal"
                "png"                   ;disables many problematic image tests
                "test_figure_legend_outside"
                "test_ft2font_get_kerning"
                "test_normal_axes"
                "test_get_tightbbox_polar")
               " and not "))
      #:phases
      #~(modify-phases %standard-phases
          (add-before 'build 'pretend-version
            ;; The version string is usually derived via setuptools-scm, but
            ;; without the git metadata available, the version string is set to
            ;; '0.0.0'.
            (lambda _
              (setenv "SETUPTOOLS_SCM_PRETEND_VERSION" #$version)))
          (add-after 'unpack 'patch-commands
            (lambda _
              (substitute* "lib/matplotlib/tests/test_animation.py"
                (("/bin/sh") (which "sh")))))
          (add-after 'unpack 'patch-dlopen
            (lambda* (#:key inputs #:allow-other-keys)
              (substitute* "src/_c_internal_utils.cpp"
                (("libX11.so.6")
                 (search-input-file inputs "lib/libX11.so.6")))))
          (add-before 'build 'configure-environment
            (lambda* (#:key inputs #:allow-other-keys)
              ;; Fix rounding errors when using the x87 FPU.
              (when (string-prefix? "i686" #$(%current-system))
                (setenv "CFLAGS" "-ffloat-store"))))
          (add-before 'check 'chdir
            (lambda _
              ;; Step out of the source directory to avoid interference.
              (chdir "/tmp"))))))
    (propagated-inputs
     (list python-contourpy
           python-cycler
           python-dateutil
           python-fonttools-minimal
           python-importlib-resources
           python-kiwisolver
           python-numpy
           python-packaging
           python-pillow
           python-pyparsing
           `(,python "tk")))
    (inputs
     (list cairo
           freetype
           glib
           libpng
           qhull
           tcl
           tk))
    (native-inputs
     (list meson-python
           pkg-config
           pybind11
           python-pypa-build
           python-pytest
           python-pytest-timeout
           python-pytest-xdist
           python-setuptools
           python-setuptools-scm
           python-wheel))
    (home-page "https://matplotlib.org/")
    (synopsis "2D plotting library for Python")
    (description
     "Matplotlib is a Python 2D plotting library which produces publication
quality figures in a variety of hardcopy formats and interactive environments
across platforms.  Matplotlib can be used in Python scripts, the python and
ipython shell, web application servers, and six graphical user interface
toolkits.")
    (license license:psfl)))
```

**Comment by QuLogic** [comment]
The `text-overhaul` branch is now merged to `main`.
